### PR TITLE
remove all generic interfaces that put Span<T> in generic arg

### LINKED ISSUE #838: Remove IEnumerable<T> from Spans.
Since Span<T> will be stack-only, it cannot implement this interface.

### PR DESCRIPTION
**Pull request by adamsitnik**
@KrzysztofCwalina I have removed it also from Utf8String since it's based on Span as well

### CODE REVIEW

**Comment by davidfowl** [comment]
We have to remove all interfaces that put `Span<T>` in the generic arg.

**Comment by adamsitnik** [comment]
@davidfowl good point!

I have removed these interfactes + `override bool Equals(object obj)` implementation that was allowing to box spans.

File: src/System.Slices/System/ReadOnlySpan.cs [hunk 1]
```diff
@@ -353,27 +349,14 @@ public override int GetHashCode()
...
-            return ReferenceEquals(other);
-        }
+        public bool Equals(ReadOnlySpan<T> other) => ReferenceEquals(other);
 
-        public bool Equals(Span<T> other)
-        {
-            return other.StructuralEquals(Object, Offset, Length);
-        }
-
-        public override bool Equals(object obj)
```
Comment (KrzysztofCwalina) [inline src/System.Slices/System/ReadOnlySpan.cs]: Why is this removed?

Comment (benaadams) [inline src/System.Slices/System/ReadOnlySpan.cs]: Should it just return false? Assume `ReadOnlySpan<T>` can't be boxed to object?

Comment (KrzysztofCwalina) [inline src/System.Slices/System/ReadOnlySpan.cs]: or throw till we have boxing disabled.

Comment (adamsitnik) [inline src/System.Slices/System/ReadOnlySpan.cs]: @KrzysztofCwalina I added throwing override for `bool Equals(object obj)`

File: src/System.Slices/System/Contract.cs [hunk 2]
```diff
@@ -83,6 +83,8 @@ public static void RequiresInInclusiveRange(uint start, uint length, uint existi
             }
         }
 
+        internal static InvalidOperationException InvalidOperationExceptionForBoxingSpans() => new InvalidOperationException("Spans must not be boxed");
```
Comment (omariom) [inline src/System.Slices/System/Contract.cs]: What are the benefits of having it cached?

Comment (omariom) [inline src/System.Slices/System/Contract.cs]: Oh.. it is not a field.. Ignore :)

File: src/System.Slices/System/ReadOnlySpan.cs (hunk 1 above)
Comment (KrzysztofCwalina) [inline src/System.Slices/System/ReadOnlySpan.cs]: Great. thanks! ... and now the tests show all the bugs :-)

Comment (adamsitnik) [inline src/System.Slices/System/ReadOnlySpan.cs]: @KrzysztofCwalina ups.. Sorry for that! Corrected!
